crates/server/src/admin/utils/auth.rs

```diff
@@ -283,6 +284,93 @@ pub fn validate_root_key_exists(
+                            message: "Wallet address not present".to_string(),
+                        });
+                    }
+                };
+                // Check if the wallet_address has a NEAR account key from DB
+                let near_keys: String = match has_near_account_root_key(store, wallet_address) {
+                    Ok(keys) if keys.is_empty() => {
+                        return Err(ApiError {
+                            status_code: StatusCode::BAD_REQUEST,
+                            message: "Root key does not exist".into(),
+                        });
+                    }
+                    Ok(keys) => keys,
+                    Err(err) => {
+                        info!("Error checking if near client key exists: {}", err);
+                        return Err(ApiError {
+                            status_code: StatusCode::INTERNAL_SERVER_ERROR,
+                            message: err.to_string(),
+                        });
+                    }
+                };
+
+                // Extract wallet_address as a &str
+                let wallet_address = match req.wallet_metadata.wallet_address.as_deref() {
+                    Some(address) => address,
+                    None => {
+                        return Err(ApiError {
+                            status_code: StatusCode::BAD_REQUEST,
+                            message: "Wallet address not present".to_string(),
+                        });
+                    }
+                };
+
+                // Get network_type and from it use correct rpc_url
+                let rpc_url = match network_id {
+                    NearNetworkId::Mainnet => "https://rpc.mainnet.near.org",
+                    NearNetworkId::Testnet => "https://rpc.testnet.near.org",
+                    _ => {
+                        // Handle the case where the network ID is unknown or not handled
+                        return Err(ApiError {
+                            status_code: StatusCode::BAD_REQUEST,
+                            message: "Unknown NEAR network ID".into(),
+                        });
+                    }
+                };
+
+                // Check if the given public key is from the given NEAR account
+                if !has_near_key(&req.wallet_metadata.verifying_key, wallet_address, rpc_url)
+                    .await?
+                {
+                    return Err(ApiError {
+                        status_code: StatusCode::BAD_REQUEST,
+                        message: format!(
+                            "Provided public key does not belong to account {:?}",
+                            wallet_address
+                        )
+                        .into(),
+                    });
+                }
+                // Check if the wallet_address has a NEAR account key from DB
+                match has_near_key(&near_keys, wallet_address, rpc_url).await? {
```

Comment:
Matching on a boolean is not generally considered idiomatic, and may be clearer as an `if` 🙂